Add `combine` to compose multiple evaluations

lvwerra opened the pull request:
This PR adds the `EvaluationEnsemble` class that lets users combine several `EvaluationModule`s into a single, compact object that behaves like a single module. Here's how it works:

```py
from evaluate import EvaluationEnsemble

clf_metrics = EvaluationEnsemble(["accuracy", "f1", "precision","recall"])
clf_metrics.compute(predictions=[0,1], references=[1,1])

{
    'accuracy': 0.5, 
    'f1': 0.666, 
    'precision': 1.0,
    'recall': 0.5
}
```

Under the hood the `add`, `add_batch`, and `compute` function iteratively call the methods with the same name for each module.

There are a few scenarios for the returned dict:
- if there are name clashes in the score names the module name is prefixed
- if there are name clashes in the modules the prefixes are additionally enumerated
- one can force the prefix with `force_prefix=True` at initialization
- for the name the module names are used except if the modules are passed as a dict in which case the keys are used.

Here an example to illustrate forced prefixing with custom names:
```py
bleu_metrics = EvaluationEnsemble({
    "bleu_simple": evaluate.load("bleu"),
    "bleu_complex": evaluate.load("bleu")}
)

{
 'bleu_simple_bleu': 0.0,
 'bleu_simple_precisions': [0.75, 0.5, 0.0, 0.0],
 'bleu_simple_brevity_penalty': 1.0,
 'bleu_simple_length_ratio': 1.0,
 'bleu_simple_translation_length': 4,
 'bleu_simple_reference_length': 4,
 'bleu_complex_bleu': 0.0,
 'bleu_complex_precisions': [0.75, 0.5, 0.0, 0.0],
 'bleu_complex_brevity_penalty': 1.0,
 'bleu_complex_length_ratio': 1.0,
 'bleu_complex_translation_length': 4,
 'bleu_complex_reference_length': 4
}
```
This should be especially useful if the same metric with different configs is used in one ensemble.

I am not married to the name `EvaluationEnsemble` so if there are any better suggestions I am happy to hear them.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/evaluate/pr_150). All of your documentation changes will be reflected on that endpoint.

douwekiela commented:
Tbh I'm not sold on the name - won't people get confused about this being a different kind of "ensemble"? I'd call it `AggregatedEvaluation` or `EvaluationAggregation`?

lhoestq commented:
I suggested `load_metrics(names: List[str])` and `combine_metrics(metrics: List[Metric])` here: https://github.com/huggingface/evaluate/issues/8#issuecomment-1093271835

Btw I think EvaluationModule is not a good name, I did not approve https://github.com/huggingface/evaluate/pull/42 - I'd like to discuss this more

lvwerra commented:
Yes, I am also not totally happy with the name, yet. The reason I didn't go for `combine_metrics` or `load_metrics` is that it is again specific to metrics whereas we have Metrics, Comparisons, and Measurements. We could go for `combine` but I feel it might be a bit too general to understand its purpose intuitively or maybe `combine_evaluations`. I also like `AggregatedEvaluation`.

Regarding `EvaluationModule` I have not a strong opinion. I feel like only a small fraction of user will create a custom module and for them it will already be populated and what's left to do is described. We can create subclasses of `EvaluationModule` for `Metric`, `Comparison`, and `Measurement` if you think that would make it more intuitive.

lhoestq commented:
> Yes, I am also not totally happy with the name, yet. The reason I didn't go for combine_metrics or load_metrics is that it is again specific to metrics whereas we have Metrics, Comparisons, and Measurements. We could go for combine but I feel it might be a bit too general to understand its purpose intuitively or maybe combine_evaluations. I also like AggregatedEvaluation.

Ok I see ! I'm fine with having `combine` at the top level to be honest ^^

> Regarding EvaluationModule I have not a strong opinion. I feel like only a small fraction of user will create a custom module and for them it will already be populated and what's left to do is described. We can create subclasses of EvaluationModule for Metric, Comparison, and Measurement if you think that would make it more intuitive.

The idea is to make it easy for anyone to read a metric/comparison/measurement script. I think it's easier to understand when we use more relevant names. Though it also depends if it is compatible your plans regarding the abstractions/separations of those types of evaluations

lvwerra commented:
Following the discussions above I added a `combine` function and renamed `EvaluationEnsemble` to `AggregateEvaluation`. Behind the scenes the `AggregateEvaluation` still does the heavy lifting and is the object returned by `combined`. 

The naming or `EvaluationModule` is addressed in #160 separately. Do you think we also need to split `AggregateEvalaution` into `AggregateMetric`, `AggregateComparison` etc.? The plan is not to expose this to the user directly by removing it from the `__init__` but they would still see it when inspecting the object returned by `combine`.

Let me know what you think.

douwekiela commented on src/evaluate/module.py:
```diff
@@ -706,3 +708,121 @@ def _enforce_nested_string_type(self, schema, obj):
+        """
+        for evaluation_module in self.evaluation_modules:
+            batch = {"predictions": prediction, "references": reference, **kwargs}
+            batch = {intput_name: batch[intput_name] for intput_name in evaluation_module._feature_names()}
```
`intput_name`? hehe

douwekiela replied:
(occurs in a few places below)

lvwerra replied:
fixed

douwekiela commented on src/evaluate/module.py:
```diff
@@ -706,3 +708,121 @@ def _enforce_nested_string_type(self, schema, obj):
+
+
+def combine(evaluation_modules, force_prefix=False):
+    """A function to comine several metrics, comparisons, and measurements into a single `AggregateEvalution` that
```
Some typos here: `combine`, `AggregateEvaluation`

douwekiela commented on src/evaluate/module.py:
```diff
@@ -706,3 +708,121 @@ def _enforce_nested_string_type(self, schema, obj):
+
+    Args:
+        evaluation_modules (``Union[list, dict]``): A list or dictionary of evaluation modules. The modules can either be passed
+            as strings or loaded `EvaluationMoudule`s. If a dictionary is passed its keys are the names used and the values the modules.
```
Typo: `Moudule`

sashavor reviewed (approved):
@douwekiela picked up on all of the typos, otherwise the logic looks good!

lhoestq commented on src/evaluate/module.py:
```diff
@@ -706,3 +708,121 @@ def _enforce_nested_string_type(self, schema, obj):
+        results_keys = list(itertools.chain.from_iterable([r.keys() for r in results]))
+        duplicate_keys = {item for item, count in collections.Counter(results_keys).items() if count > 1}
+
+        duplicate_names = [
+            item for item, count in collections.Counter(self.evaluation_module_names).items() if count > 1
+        ]
+        duplicate_counter = {name: 0 for name in duplicate_names}
```
It's also possible to simply raise an error for now, if we need more time to think about how to handle this case.

As a user it's a bit unclear for me that it would automatically use prefixes when there are name conflicts, but no prefixes if there are no conflicts. This is the kind of black magic we try to avoid usually - though I don't have a strong opinion on this very specific case, since it would be easy to explain in the docs. What do you think ?

an alternative could be to ask users to name the metrics:
```python
combine({
    "metric_1": metric_1,
    "metric_2": metric_2
})
```

lvwerra replied:
What do you think about adding a warning when a prefixes are added:
```py
f"The key '{key}' occurs in several evaluation results. Prefixing with the key with '{prefix}'."
```

The user can already use a `dict` in the current state of the PR (see last example in initial comment). What do you think?

lhoestq replied:
A warning sounds good thanks !

lhoestq reviewed (commented):
Cool ! More comments that mostly come from the Open Source experience I got from the past two years. I'm happy to share these thoughts but please ignore if they slow you down too much

lvwerra commented on src/evaluate/module.py:
```diff
@@ -706,3 +708,121 @@ def _enforce_nested_string_type(self, schema, obj):
+        return merged_results
+
+
+def combine(evaluation_modules, force_prefix=False):
```
Actually also like `CombinedEvaluations` better. If @douwekiela is also happy with that I'll rename to this.

douwekiela reviewed (approved):
LGTM

lvwerra commented on tests/test_metric.py:
```diff
@@ -612,3 +628,82 @@ def test_metric_with_non_standard_feature_names_compute(tmp_path):
+                expected_results[f"{dummy_metric.name}_{i}_{k}"] = dummy_result[k]
+        self.assertDictEqual(expected_results, combined_evaluation.compute(predictions=preds, references=refs))
+
+    def test_two_modules_with_same_score_name(self):
```
I think that is covered in `test_duplicate_module`, no?

lvwerra replied:
I merge this now, if that's not the case I can add more tests in another PR.
